### PR TITLE
Better error handling in OrdersClient

### LINKED ISSUE #325: The locations property in models.py breaks when order fails
On branch v2-polling-output-310, in models.py, `locations()` returns a TypeError when trying to unpack all download locations when `self.results` is a NoneType, i.e., when an order fails. Specifically, this occurs at line 380 in the following return statement:
```
return list(r[self.LOCATION_KEY] for r in self.results)
```

Below is the code used for the request, poll, and download the order, taken from docs/guide.md:
```
import asyncio
from planet import Session
from planet import OrdersClient

# Build request
request = {
  "name": "test_order",
  "products": [
    {
      "item_ids": [
        "20170614_113217_3163208_RapidEye-5"
      ],
      "item_type": "REOrthoTile",
      "product_bundle": "analytic"
    }
  ],
  "tools": [
    {
      "toar": {
        "scale_factor": 10000
      }
    },
    {
      "reproject": {
        "projection": "WSG84",
        "kernel": "cubic"
      }
    },
    {
      "tile": {
        "tile_size": 1232,
        "origin_x": -180,
        "origin_y": -90,
        "pixel_size": 2.7056277056e-05,
        "name_template": "C1232_30_30_{tilex:04d}_{tiley:04d}"
      }
    }
  ]
}    

# Polling & Downloading
from planet import reporting

async def create_poll_and_download():
    async with Session() as sess:
        cl = OrdersClient(sess)
        with reporting.StateBar(state='creating') as bar:
            # create order
            order = await cl.create_order(request)
            bar.update(state='created', order_id=order.id)

            # poll
            await cl.poll(order.id, report=bar.update)

        # download
        await cl.download_order(order.id)

asyncio.run(create_poll_and_download())
```

Below is the traceback for the aforementioned error:

```
TypeError                                 Traceback (most recent call last)
<ipython-input-12-401800b0f64e> in <module>
     13         await cl.download_order(order.id)
     14
---> 15 asyncio.run(create_poll_and_download())

/usr/local/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/runners.py in run(main, debug)
     42         if debug is not None:
     43             loop.set_debug(debug)
---> 44         return loop.run_until_complete(main)
     45     finally:
     46         try:

/usr/local/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py in run_until_complete(self, future)
    640             raise RuntimeError('Event loop stopped before Future completed.')
    641
--> 642         return future.result()
    643
    644     def stop(self):

<ipython-input-12-401800b0f64e> in create_poll_and_download()
     11
     12         # download
---> 13         await cl.download_order(order.id)
     14
     15 asyncio.run(create_poll_and_download())

~/code/repos/planet-client-python/planet/clients/orders.py in download_order(self, order_id, directory, overwrite, progress_bar)
    332         """
    333         order = await self.get_order(order_id)
--> 334         locations = order.locations
    335         LOGGER.info(
    336             f'downloading {len(locations)} assets from order {order_id}'

~/code/repos/planet-client-python/planet/models.py in locations(self)
    378         :rtype: list of str
    379         '''
--> 380         return list(r[self.LOCATION_KEY] for r in self.results)
    381
    382     @property

TypeError: 'NoneType' object is not iterable
```

**Summary**
What's happening is that when an order fails, `self.results` will be equal to `None`, which is not iterable, which was inherited from `order.locations` in orders.py (~line 334).

**Proposal** 
1. Change the order in the guide (`docs/guide.md`) with an order that doesn’t fail.
2. Change the property `locations()` within models.py to be protected agains't failed scenarios. 
For example:
```
@property
def locations(self):
    '''Download locations for order results.

    :return: download locations in order
    :rtype: list of str
    '''
    try:
        return list(r[self.LOCATION_KEY] for r in self.results)
    except TypeError:
        return []
```

### PR DESCRIPTION
**Pull request by kevinlacaille**
Added a try/except clause to `planet.clients.orders.OrdersClient._get_order_locations()` to protect against failed scenarios, where `list(r['location'] for r in results if r)` would return an empty list.

Closes #325

### CODE REVIEW

File: planet/clients/orders.py [hunk 1]
```diff
@@ -319,7 +319,10 @@ async def download_order(self,
     def _get_order_locations(order):
         links = order['_links']
         results = links.get('results', None)
-        return list(r['location'] for r in results if r)
+        try:
```
[the remark this answers is not in the document]
Comment (kevinlacaille) [reply planet/clients/orders.py:320]: Yeah, this would probably do it! However, what do you think about raising a ClientError exception if no links are found (i.e., L323)?

[the remark this answers is not in the document]
Comment (jreiberkyle) [reply planet/clients/orders.py:320]: hmmm... that is a great question. Let's think about usage here. For `_get_order_locations` it doesn't really seem to be an error scenario if there are no results - there just are no locations either (hence the empty list being returned). Maybe `download_orders` would be concerned about that but if locations is an empty list, it doesn't download any files and `filenames` is an empty list. So, it returns an empty list if no links are found. Is it considered an error if a user runs `download_order` and the order has no links, so no files were downloaded? Should that be a scenario that causes the program to halt or requires catching the exception? I'm quite open to dialogue. My initial thinking is that it shouldn't raise an exception since the user might not mind an order being empty, but it should raise a warning. Maybe that's the least helpful approach, but it's also the least heavy-handed. What are your thoughts?

[the remark this answers is not in the document]
Comment (kevinlacaille) [reply planet/clients/orders.py:320]: That's a good point!

Let's not raise an exception, because you're right, it should still continue to run even if no locations are provided. So, instead of raising an exception I'll return an empty list and warn the user via `LOGGER.warning()` that there were no locations found and that nothing will be downloaded.

[the remark this answers is not in the document]
Comment (jreiberkyle) [reply planet/clients/orders.py:320]: sounds great! and if we find we want to change our minds later, it will be easy to do so. Thanks for going through that thought exercise with me

File: planet/clients/orders.py [hunk 2]
```diff
@@ -318,8 +316,14 @@ async def download_order(self,
     @staticmethod
     def _get_order_locations(order):
         links = order['_links']
-        results = links.get('results', None)
-        return list(r['location'] for r in results if r)
+        results = links.get('results', [])
+        try:
+            return list(r['location'] for r in results if r)
+        except TypeError:
```
Comment (jreiberkyle) [inline planet/clients/orders.py:322]: wow! so this still throws a TypeError even with the change to line 319 and the `...if r` check? That's a surprise to me!

File: tests/integration/test_orders_api.py [hunk 3]
```diff
@@ -645,6 +645,37 @@ async def test_download_order_state(tmpdir, order_description, oid, session):
...
+@respx.mock
+@pytest.mark.asyncio
+async def test_download_order_without_order_links(tmpdir,
+                                                  order_description,
+                                                  oid,
+                                                  session):
+    dl_url1 = TEST_DOWNLOAD_URL + '/1?token=IAmAToken'
+
+    order_description['state'] = 'success'
+    order_description['_links']['results'] = None
```
Comment (jreiberkyle) [inline tests/integration/test_orders_api.py:657]: Just a conversation: It is kind of wacky that the order is in a success state and yet has no results to download. I wonder if we ever actually see that scenario.

**Review by jreiberkyle — APPROVED** [review]
overall looks great. I had some discoveries and conversations I'd love to get a chance to discuss with you but they shouldn't be a blocker to a merge.

File: planet/clients/orders.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevinlacaille) [reply planet/clients/orders.py:320]: Alrighty, I'll leave it as a warning for now and we can change our mind later, if we wish :)

File: tests/integration/test_orders_api.py (hunk 3 above)
Comment (kevinlacaille) [reply tests/integration/test_orders_api.py:657]: Right, totally agree, but for some reason that bubbled up in #325, so maybe just a weird edge case?
